File: sdk/typescript/src/utils/input-processor.ts
import { lookup } from "node:dns/promises";
import * as ipaddr from "ipaddr.js";
import { getDocumentProxy } from "unpdf";
import type { GuardInput, ProcessedInput } from "../types.js";

/**
 * Image MIME types supported by vision models
 */
const IMAGE_MIME_TYPES = [
  "image/png",
  "image/jpeg",
  "image/jpg",
  "image/gif",
  "image/webp",
];

/**
 * PDF MIME types
 */
const PDF_MIME_TYPES = ["application/pdf"];

/**
 * Text MIME types that can be processed directly
 */
const TEXT_MIME_TYPES = [
  "text/plain",
  "text/html",
  "text/css",
  "text/javascript",
  "text/csv",
  "text/xml",
  "application/json",
  "application/xml",
];

/**
 * Check if a string looks like a URL
 */
function isUrlString(input: string): boolean {
  return input.startsWith("http://") || input.startsWith("https://");
}

/**
 * Check if an IP address is private/internal using ipaddr.js
 * Supports both IPv4 and IPv6 ranges
 */
function isPrivateIpAddress(ip: string): boolean {
  try {
    const addr = ipaddr.process(ip);
    
    if (addr.kind() === "ipv4") {
      const ipv4 = addr as ipaddr.IPv4;
      
      // Check IPv4 private ranges
      return (
        ipv4.range() === "private" || // 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16
        ipv4.range() === "loopback" || // 127.0.0.0/8
        ipv4.range() === "linkLocal" || // 169.254.0.0/16
        ipv4.range() === "multicast" // 224.0.0.0/4
      );
    } else if (addr.kind() === "ipv6") {
      const ipv6 = addr as ipaddr.IPv6;
      
      // Check for IPv4-mapped IPv6 addresses (::ffff:127.0.0.1)
      if (ipv6.isIPv4MappedAddress()) {
        const ipv4Mapped = ipv6.toIPv4Address();
        return isPrivateIpAddress(ipv4Mapped.toString());
      }
      
      // Check IPv6 private ranges
      const range = ipv6.range();
      return (
        range === "uniqueLocal" || // fc00::/7 (ULA)
        range === "linkLocal" || // fe80::/10
        range === "loopback" || // ::1
        range === "multicast" // ff00::/8
      );
    }
    
    return false;
  } catch {
    // If IP parsing fails, treat as private (fail-safe)
    return true;
  }
}

/**
 * Check if a hostname resolves to a private/internal IP address.
 * 
 * This function performs DNS resolution to prevent SSRF attacks where
 * a hostname like "attacker.com" resolves to 127.0.0.1.
 * 
 * Blocks:
 * - IPv4: 127.0.0.0/8, 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16, 169.254.0.0/16, 224.0.0.0/4
 * - IPv6: ::1, fc00::/7, fe80::/10, ::ffff:127.0.0.0/104, ff00::/8
 * - localhost hostnames
 * 
 * If DNS resolution fails, treats as private (fail-safe security).
 */
async function isPrivateIp(hostname: string): Promise<boolean> {
  const lowerHostname = hostname.toLowerCase();

  // Check for localhost hostname first (common case, no DNS needed)
  if (
    lowerHostname === "localhost" ||
    lowerHostname === "localhost.localdomain" ||
    lowerHostname === "local" ||
    lowerHostname === "127.0.0.1" ||
    lowerHostname === "::1"
  ) {
    return true;
  }

  // Remove brackets from IPv6 addresses if present
  const cleanHostname = hostname.replace(/^\[|\]$/g, "");

  // Check if it's already a direct IP address (IPv4 or IPv6)
  try {
    // Try parsing as IP first (faster path)
    if (ipaddr.isValid(cleanHostname)) {
      return isPrivateIpAddress(cleanHostname);
    }
  } catch {
    // Not a direct IP, continue to DNS resolution
  }

  // Perform DNS resolution to get actual IP address
  // This prevents SSRF where hostname resolves to private IP
  try {
    const result = await lookup(cleanHostname, { all: false });
    const resolvedIp = result.address;
    
    // Check the resolved IP address
    return isPrivateIpAddress(resolvedIp);
  } catch (error) {
    // DNS resolution failed - treat as private (fail-safe)
    // This prevents bypasses where DNS fails but might resolve to private IP
    return true;
  }
}

/**
 * Validate URL for security concerns.
 *
 * Throws Error with descriptive message if URL is invalid or unsafe.
 *
 * Checks:
 * - URL format is valid
 * - Protocol is http or https only
 * - Hostname is not empty
 * - No private/internal IP addresses (with DNS resolution)
 * - No localhost access
 * - No file:// protocol
 * - URL length is reasonable (max 2048 characters)
 */
async function validateUrl(url: string): Promise<void> {
  const MAX_URL_LENGTH = 2048;

  // Check URL length
  if (url.length > MAX_URL_LENGTH) {
    throw new Error(
      `Invalid URL: URL exceeds maximum length of ${MAX_URL_LENGTH} characters`
    );
  }

  // Parse URL
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch (e) {
    throw new Error(`Invalid URL: malformed URL format - ${String(e)}`);
  }

  // Check protocol
  const protocol = parsed.protocol.toLowerCase();
  if (protocol !== "http:" && protocol !== "https:") {
    if (protocol === "file:") {
      throw new Error("Invalid URL: file:// protocol is not allowed");
    }
    throw new Error(
      `Invalid URL: protocol must be http or https, got ${protocol}`
    );
  }

  // Check hostname
  const hostname = parsed.hostname;
  if (!hostname) {
    throw new Error("Invalid URL: hostname is required");
  }

  // Check for private/internal IP addresses (with DNS resolution)
  const isPrivate = await isPrivateIp(hostname);
  if (isPrivate) {
    const lowerHostname = hostname.toLowerCase();
    if (
      lowerHostname === "localhost" ||
      hostname === "127.0.0.1" ||
      hostname === "::1" ||
      hostname.startsWith("127.")
    ) {
      throw new Error("Invalid URL: localhost access is not allowed");
    }
    throw new Error(
      "Invalid URL: private/internal IP addresses are not allowed"
    );
  }
}

/**
 * Check if MIME type is an image
 */
function isImageMimeType(mimeType: string): boolean {
  return IMAGE_MIME_TYPES.some((type) => mimeType.startsWith(type));
}

/**
 * Check if MIME type is text-based
 */
function isTextMimeType(mimeType: string): boolean {
  return (
    TEXT_MIME_TYPES.some((type) => mimeType.startsWith(type)) ||
    mimeType.startsWith("text/")
  );
}

/**
 * Check if MIME type is PDF
 */
function isPdfMimeType(mimeType: string): boolean {
  return PDF_MIME_TYPES.some((type) => mimeType.startsWith(type));
}

/**
 * Get MIME type from URL based on extension
 */
function getMimeTypeFromUrl(url: string): string | null {
  const pathname = new URL(url).pathname.toLowerCase();
  if (pathname.endsWith(".png")) return "image/png";
  if (pathname.endsWith(".jpg") || pathname.endsWith(".jpeg"))
    return "image/jpeg";
  if (pathname.endsWith(".gif")) return "image/gif";
  if (pathname.endsWith(".webp")) return "image/webp";
  if (pathname.endsWith(".pdf")) return "application/pdf";
  if (pathname.endsWith(".txt")) return "text/plain";
  if (pathname.endsWith(".html") || pathname.endsWith(".htm"))
    return "text/html";
  if (pathname.endsWith(".json")) return "application/json";
  if (pathname.endsWith(".xml")) return "application/xml";
  if (pathname.endsWith(".csv")) return "text/csv";
  return null;
}

/**
 * Convert ArrayBuffer to base64 string
 */
function arrayBufferToBase64(buffer: ArrayBuffer): string {
  const bytes = new Uint8Array(buffer);
  let binary = "";
  for (let i = 0; i < bytes.byteLength; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}

/**
 * Extract text from each page of a PDF
 * Returns an array of strings, one per page
 */
async function extractPdfPages(buffer: ArrayBuffer): Promise<string[]> {
  const pdf = await getDocumentProxy(new Uint8Array(buffer));
  const pages: string[] = [];

  for (let i = 1; i <= pdf.numPages; i++) {
    const page = await pdf.getPage(i);
    const textContent = await page.getTextContent();
    const pageText = textContent.items
      .map((item) => {
        // TextItem has 'str' property, TextMarkedContent does not
        if ("str" in item) {
          return item.str;
        }
        return "";
      })
      .join(" ");
    pages.push(pageText);
  }

  return pages;
}

/**
 * Fetch content from a URL and return as ProcessedInput
 */
async function fetchUrl(url: string): Promise<ProcessedInput> {
  // Validate URL for security before making any network requests
  await validateUrl(url);

  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(
      `Failed to fetch URL: ${response.status} ${response.statusText}`
    );
  }

  // Get content type from response or infer from URL
  let contentType =
    response.headers.get("content-type")?.split(";")[0].trim() || "";

  // If no content type, try to infer from URL
  if (!contentType) {
    contentType = getMimeTypeFromUrl(url) || "text/plain";
  }

  if (isPdfMimeType(contentType)) {
    const buffer = await response.arrayBuffer();
    const pages = await extractPdfPages(buffer);
    return {
      type: "pdf",
      pages,
      mimeType: contentType,
    };
  }

  if (isImageMimeType(contentType)) {
    const buffer = await response.arrayBuffer();
    const base64 = arrayBufferToBase64(buffer);
    return {
      type: "image",
      imageBase64: base64,
      mimeType: contentType,
    };
  }

  if (isTextMimeType(contentType)) {
    const text = await response.text();
    return {
      type: "text",
      text,
      mimeType: contentType,
    };
  }

  // For other content types, try to read as text
  // This handles cases where server doesn't send correct content-type
  try {
    const text = await response.text();
    return {
      type: "text",
      text,
      mimeType: contentType,
    };
  } catch {
    throw new Error(`Unsupported content type: ${contentType}`);
  }
}

/**
 * Process a Blob and return as ProcessedInput
 */
async function processBlob(blob: Blob): Promise<ProcessedInput> {
  const mimeType = blob.type || "application/octet-stream";

  if (isPdfMimeType(mimeType)) {
    const buffer = await blob.arrayBuffer();
    const pages = await extractPdfPages(buffer);
    return {
      type: "pdf",
      pages,
      mimeType,
    };
  }

  if (isImageMimeType(mimeType)) {
    const buffer = await blob.arrayBuffer();
    const base64 = arrayBufferToBase64(buffer);
    return {
      type: "image",
      imageBase64: base64,
      mimeType,
    };
  }

  if (isTextMimeType(mimeType)) {
    const text = await blob.text();
    return {
      type: "text",
      text,
      mimeType,
    };
  }

  // Try to read as text for unknown types
  try {
    const text = await blob.text();
    return {
      type: "text",
      text,
      mimeType,
    };
  } catch {
    throw new Error(`Unsupported blob type: ${mimeType}`);
  }
}

/**
 * Process guard input and return normalized ProcessedInput
 *
 * Auto-detection:
 * - string starting with http:// or https:// → fetch URL
 * - string not starting with http(s):// → plain text
 * - Blob/File → process based on MIME type
 * - URL object → fetch URL
 */
export async function processInput(input: GuardInput): Promise<ProcessedInput> {
  // Handle URL object
  if (input instanceof URL) {
    return fetchUrl(input.toString());
  }

  // Handle Blob (includes File)
  if (typeof Blob !== "undefined" && input instanceof Blob) {
    return processBlob(input);
  }

  // Handle string
  if (typeof input === "string") {
    // Check if it's a URL
    if (isUrlString(input)) {
      return fetchUrl(input);
    }

    // Plain text
    return {
      type: "text",
      text: input,
      mimeType: "text/plain",
    };
  }

  throw new Error("Invalid input type");
}

/**
 * Check if a model supports vision (multimodal) input
 */
export function isVisionModel(model: string): boolean {
  // OpenAI vision models (gpt-4o, gpt-4o-mini, gpt-4-turbo, gpt-4.1 all support vision)
  if (
    model.includes("gpt-4o") ||
    model.includes("gpt-4-turbo") ||
    model.includes("gpt-4.1") ||
    model.includes("gpt-5")
  ) {
    return true;
  }
  // Anthropic Claude 3+ models support vision
  if (
    model.includes("claude-3") ||
    model.includes("claude-sonnet-4") ||
    model.includes("claude-opus-4") ||
    model.includes("claude-haiku-4")
  ) {
    return true;
  }
  // Google Gemini models support vision
  if (model.includes("gemini")) {
    return true;
  }
  // Vercel models with vision
  if (model.includes("grok-2-vision") || model.includes("pixtral")) {
    return true;
  }
  // OpenRouter vision models
  if (
    model.includes("vision") ||
    model.includes("-vl-") ||
    model.includes("pixtral")
  ) {
    return true;
  }
  return false;
}

File: sdk/python/src/safety_agent/utils/input_processor.py
"""
Input processor for handling various input types (text, URLs, images, PDFs)
"""

import base64
import ipaddress
import socket
from urllib.parse import urlparse

import httpx

from ..types import GuardInput, ProcessedInput

# Image MIME types supported by vision models
IMAGE_MIME_TYPES = [
    "image/png",
    "image/jpeg",
    "image/jpg",
    "image/gif",
    "image/webp",
]

# PDF MIME types
PDF_MIME_TYPES = ["application/pdf"]

# Text MIME types that can be processed directly
TEXT_MIME_TYPES = [
    "text/plain",
    "text/html",
    "text/css",
    "text/javascript",
    "text/csv",
    "text/xml",
    "application/json",
    "application/xml",
]


def _is_url_string(input_str: str) -> bool:
    """Check if a string looks like a URL."""
    return input_str.startswith("http://") or input_str.startswith("https://")


def _is_private_ip(hostname: str) -> bool:
    """
    Check if a hostname resolves to a private/internal IP address.
    
    Blocks:
    - localhost and 127.0.0.0/8 (loopback)
    - 10.0.0.0/8 (private)
    - 172.16.0.0/12 (private)
    - 192.168.0.0/16 (private)
    - localhost hostname
    """
    # Check for localhost hostname
    if hostname.lower() in ("localhost", "localhost.localdomain", "local"):
        return True
    
    # Try to resolve hostname to IP address
    try:
        # Get the first IP address from hostname
        ip_str = socket.gethostbyname(hostname)
        ip = ipaddress.ip_address(ip_str)
        
        # Check if it's a private IP
        return ip.is_loopback or ip.is_private or ip.is_link_local or ip.is_multicast
    except (socket.gaierror, socket.herror, ValueError):
        # If hostname can't be resolved or is not a valid IP, check if it's a direct IP
        try:
            ip = ipaddress.ip_address(hostname)
            return ip.is_loopback or ip.is_private or ip.is_link_local or ip.is_multicast
        except ValueError:
            # Not a valid IP address, but we'll allow it (could be a domain)
            # The actual connection will fail if it's invalid
            return False


def _validate_url(url: str) -> None:
    """
    Validate URL for security concerns.
    
    Raises ValueError with descriptive message if URL is invalid or unsafe.
    
    Checks:
    - URL format is valid
    - Protocol is http or https only
    - Hostname is not empty
    - No private/internal IP addresses
    - No localhost access
    - No file:// protocol
    - URL length is reasonable (max 2048 characters)
    """
    # Check URL length
    MAX_URL_LENGTH = 2048
    if len(url) > MAX_URL_LENGTH:
        raise ValueError(f"Invalid URL: URL exceeds maximum length of {MAX_URL_LENGTH} characters")
    
    # Parse URL
    try:
        parsed = urlparse(url)
    except Exception as e:
        raise ValueError(f"Invalid URL: malformed URL format - {str(e)}")
    
    # Check protocol
    protocol = parsed.scheme.lower()
    if protocol not in ("http", "https"):
        if protocol == "file":
            raise ValueError("Invalid URL: file:// protocol is not allowed")
        raise ValueError(f"Invalid URL: protocol must be http or https, got {protocol}")
    
    # Check hostname
    hostname = parsed.hostname
    if not hostname:
        raise ValueError("Invalid URL: hostname is required")
    
    # Check for private/internal IP addresses
    if _is_private_ip(hostname):
        if hostname.lower() in ("localhost", "127.0.0.1", "::1") or hostname.startswith("127."):
            raise ValueError("Invalid URL: localhost access is not allowed")
        raise ValueError("Invalid URL: private/internal IP addresses are not allowed")


def _is_image_mime_type(mime_type: str) -> bool:
    """Check if MIME type is an image."""
    return any(mime_type.startswith(t) for t in IMAGE_MIME_TYPES)


def _is_text_mime_type(mime_type: str) -> bool:
    """Check if MIME type is text-based."""
    return (
        any(mime_type.startswith(t) for t in TEXT_MIME_TYPES)
        or mime_type.startswith("text/")
    )


def _is_pdf_mime_type(mime_type: str) -> bool:
    """Check if MIME type is PDF."""
    return any(mime_type.startswith(t) for t in PDF_MIME_TYPES)


def _get_mime_type_from_url(url: str) -> str | None:
    """Get MIME type from URL based on extension."""
    parsed = urlparse(url)
    pathname = parsed.path.lower()

    if pathname.endswith(".png"):
        return "image/png"
    if pathname.endswith(".jpg") or pathname.endswith(".jpeg"):
        return "image/jpeg"
    if pathname.endswith(".gif"):
        return "image/gif"
    if pathname.endswith(".webp"):
        return "image/webp"
    if pathname.endswith(".pdf"):
        return "application/pdf"
    if pathname.endswith(".txt"):
        return "text/plain"
    if pathname.endswith(".html") or pathname.endswith(".htm"):
        return "text/html"
    if pathname.endswith(".json"):
        return "application/json"
    if pathname.endswith(".xml"):
        return "application/xml"
    if pathname.endswith(".csv"):
        return "text/csv"

    return None


def _bytes_to_base64(data: bytes) -> str:
    """Convert bytes to base64 string."""
    return base64.b64encode(data).decode("utf-8")


async def _extract_pdf_pages(data: bytes) -> list[str]:
    """Extract text from each page of a PDF."""
    from pypdf import PdfReader
    from io import BytesIO

    reader = PdfReader(BytesIO(data))
    pages = []

    for page in reader.pages:
        text = page.extract_text() or ""
        pages.append(text)

    return pages


async def _fetch_url(url: str) -> ProcessedInput:
    """Fetch content from a URL and return as ProcessedInput."""
    # Validate URL for security before making any network requests
    _validate_url(url)
    
    async with httpx.AsyncClient() as client:
        response = await client.get(url, follow_redirects=True, timeout=30.0)

        if response.status_code != 200:
            raise RuntimeError(
                f"Failed to fetch URL: {response.status_code} {response.reason_phrase}"
            )

        # Get content type from response or infer from URL
        content_type = response.headers.get("content-type", "").split(";")[0].strip()

        if not content_type:
            content_type = _get_mime_type_from_url(url) or "text/plain"

        data = response.content

        if _is_pdf_mime_type(content_type):
            pages = await _extract_pdf_pages(data)
            return ProcessedInput(
                type="pdf",
                pages=pages,
                mime_type=content_type,
            )

        if _is_image_mime_type(content_type):
            return ProcessedInput(
                type="image",
                image_base64=_bytes_to_base64(data),
                mime_type=content_type,
            )

        if _is_text_mime_type(content_type):
            return ProcessedInput(
                type="text",
                text=data.decode("utf-8"),
                mime_type=content_type,
            )

        # For other content types, try to read as text
        try:
            return ProcessedInput(
                type="text",
                text=data.decode("utf-8"),
                mime_type=content_type,
            )
        except UnicodeDecodeError:
            raise RuntimeError(f"Unsupported content type: {content_type}")


async def _process_bytes(data: bytes, mime_type: str | None = None) -> ProcessedInput:
    """Process bytes and return as ProcessedInput."""
    # Try to detect content type from magic bytes
    detected_mime = mime_type or "application/octet-stream"

    # Check for PDF magic bytes
    if data[:4] == b"%PDF":
        detected_mime = "application/pdf"
    # Check for PNG magic bytes
    elif data[:8] == b"\x89PNG\r\n\x1a\n":
        detected_mime = "image/png"
    # Check for JPEG magic bytes
    elif data[:2] == b"\xff\xd8":
        detected_mime = "image/jpeg"
    # Check for GIF magic bytes
    elif data[:6] in (b"GIF87a", b"GIF89a"):
        detected_mime = "image/gif"
    # Check for WebP magic bytes
    elif data[:4] == b"RIFF" and data[8:12] == b"WEBP":
        detected_mime = "image/webp"

    if _is_pdf_mime_type(detected_mime):
        pages = await _extract_pdf_pages(data)
        return ProcessedInput(
            type="pdf",
            pages=pages,
            mime_type=detected_mime,
        )

    if _is_image_mime_type(detected_mime):
        return ProcessedInput(
            type="image",
            image_base64=_bytes_to_base64(data),
            mime_type=detected_mime,
        )

    if _is_text_mime_type(detected_mime):
        return ProcessedInput(
            type="text",
            text=data.decode("utf-8"),
            mime_type=detected_mime,
        )

    # Try to read as text for unknown types
    try:
        return ProcessedInput(
            type="text",
            text=data.decode("utf-8"),
            mime_type=detected_mime,
        )
    except UnicodeDecodeError:
        raise RuntimeError(f"Unsupported content type: {detected_mime}")


async def process_input(input_data: GuardInput) -> ProcessedInput:
    """
    Process guard input and return normalized ProcessedInput.

    Auto-detection:
    - string starting with http:// or https:// -> fetch URL
    - string not starting with http(s):// -> plain text
    - bytes -> process based on content type detection
    """
    # Handle bytes
    if isinstance(input_data, bytes):
        return await _process_bytes(input_data)

    # Handle string
    if isinstance(input_data, str):
        # Check if it's a URL
        if _is_url_string(input_data):
            return await _fetch_url(input_data)

        # Plain text
        return ProcessedInput(
            type="text",
            text=input_data,
            mime_type="text/plain",
        )

    raise ValueError("Invalid input type")


def is_vision_model(model: str) -> bool:
    """Check if a model supports vision (multimodal) input."""
    # OpenAI vision models
    if any(x in model for x in ["gpt-4o", "gpt-4-turbo", "gpt-4.1", "gpt-5"]):
        return True

    # Anthropic Claude 3+ models support vision
    if any(x in model for x in [
        "claude-3",
        "claude-sonnet-4",
        "claude-opus-4",
        "claude-haiku-4",
    ]):
        return True

    # Google Gemini models support vision
    if "gemini" in model:
        return True

    # Vercel/OpenRouter models with vision
    if any(x in model for x in ["grok-2-vision", "pixtral", "vision", "-vl-"]):
        return True

    return False
